controllers/state_of_the_world.go

```diff
@@ -63,15 +71,122 @@ func NewPolicyMachineryController(manager ctrlruntime.Manager, client *dynamic.D
 }
 
 func buildReconciler(client *dynamic.DynamicClient) controller.ReconcileFunc {
+	authorinoCrSubscription := &controller.Subscription{
+		ReconcileFunc: NewAuthorinoCrReconciler(client).Reconcile,
+		Events: []controller.ResourceEventMatcher{
+			{Kind: ptr.To(kuadrantv1beta1.KuadrantKind), EventType: ptr.To(controller.CreateEvent)},
+			{Kind: ptr.To(kuadrantv1beta1.AuthorinoKind), EventType: ptr.To(controller.DeleteEvent)},
+		},
+	}
 	reconciler := &controller.Workflow{
 		Precondition: NewEventLogger().Log,
 		Tasks: []controller.ReconcileFunc{
 			NewTopologyFileReconciler(client, operatorNamespace).Reconcile,
+			authorinoCrSubscription.Reconcile,
 		},
 	}
 	return reconciler.Run
 }
 
+type AuthorinoCrReconciler struct {
+	Client *dynamic.DynamicClient
+}
+
+func NewAuthorinoCrReconciler(client *dynamic.DynamicClient) *AuthorinoCrReconciler {
+	return &AuthorinoCrReconciler{Client: client}
+}
+
+func (r *AuthorinoCrReconciler) Reconcile(ctx context.Context, events []controller.ResourceEvent, topology *machinery.Topology, _ error) {
+	logger := controller.LoggerFromContext(ctx).WithName("AuthorinoCrReconciler")
+	logger.Info("Reconciling Authorino Cr")
+
+	kobj := &kuadrantv1beta1.Kuadrant{}
+	for _, event := range events {
+		if event.Kind == kuadrantv1beta1.KuadrantKind && event.EventType == controller.CreateEvent {
+			kobjs := lo.FilterMap(topology.Objects().Roots(), func(item machinery.Object, _ int) (*kuadrantv1beta1.Kuadrant, bool) {
+				if item.GetName() == event.NewObject.GetName() && item.GetNamespace() == event.NewObject.GetNamespace() && item.GroupVersionKind().Kind == event.NewObject.GetObjectKind().GroupVersionKind().Kind {
+					return item.(*kuadrantv1beta1.Kuadrant), true
+				}
+				return nil, false
+			})
+			if len(kobjs) != 1 {
+				logger.Error(fmt.Errorf("muiltply kuadrant CRs found"), "unexpected behaviour may happen")
+			}
+			kobj = kobjs[0]
+			break
+		} else if event.Kind == kuadrantv1beta1.AuthorinoKind && event.EventType == controller.DeleteEvent {
+			kobjs := lo.FilterMap(topology.Objects().Roots(), func(item machinery.Object, _ int) (*kuadrantv1beta1.Kuadrant, bool) {
+				if item.GetNamespace() == event.NewObject.GetNamespace() && item.GroupVersionKind().Kind == kuadrantv1beta1.KuadrantKind.Kind {
+					return item.(*kuadrantv1beta1.Kuadrant), true
+				}
+				return nil, false
+			})
+
+			if len(kobjs) == 0 {
+				logger.Info("no possible kuadrant parent, wont create Authorino CR.")
+				return
+			}
+			if len(kobjs) != 1 {
+				logger.Error(fmt.Errorf("muiltply kuadrant CRs found"), "unexpected behaviour may happen")
+			}
+			kobj = kobjs[0]
+			if kobj.GetDeletionTimestamp() != nil {
+				logger.Info("kuadrant CR marked for deletion, wont create Authorino CR.")
+				return
+			}
+			break
+		}
+	}
```

Comment:
I had pictured this a bit different in my head:
- Find a `Kuadrant` object that is the root of the topology
- If there is none,
  - delete all `Authorino` objects that exist in the topology
  - return
- Otherwise
  - build the desired `Authorino` object for the `Kuadrant` root
  - create it if it doesn't exist as a child of the `Kuadrant` root
  - update it otherwise

Reply:
I think want is being done is mostly what you listed above. The things that are not being done are:
* deleting the Authorino CRs when the there is not kuadrant root. The cluster is handling this by the owner reference
* update if otherwise. The current implantation also does not do this. I don't think it would be right to introduction that change now.